File: crates/rivet-view/tests/view_compiler_test.rs
use noesis::{HardState, NoesisEvent, SoftWorkspace};
use rivet_types::*;
use rivet_view::{CognitiveViewCompiler, CompilationContext, RepresentationMode};

#[test]
fn test_view_compiler_all_4_modes() {
    let mut hard = HardState::new();
    let session_id = SessionId::new();
    let mut soft = SoftWorkspace::new(session_id, hard.revision);

    soft.set_focus(vec!["auth".into(), "security".into()]);
    soft.add_hypothesis("H1: OAuth token expiry mismatch");
    soft.add_unknown("U1: clock sync status");

    let claim_id = ClaimId::new();
    let evid_id = EvidenceId::new();
    let ob_id = ObligationId::new();

    hard.apply(&NoesisEvent::EvidenceRecorded {
        evidence_id: evid_id.clone(),
        source: "cargo_test".into(),
        summary: "test_auth passed".into(),
        timestamp: chrono::Utc::now(),
    });

    hard.apply(&NoesisEvent::ClaimAsserted {
        claim_id: claim_id.clone(),
        proposition: "auth middleware is threadsafe".into(),
        status: EpistemicStatus::Supported,
        evidence: vec![evid_id.clone()],
        scope: Scope::global("rivet", hard.revision),
        timestamp: chrono::Utc::now(),
    });

    hard.apply(&NoesisEvent::ObligationCreated {
        obligation_id: ob_id.clone(),
        description: "verify auth refresh route".into(),
        scope: Scope::path("rivet", "src/auth/**", hard.revision),
        timestamp: chrono::Utc::now(),
    });

    let claim2_id = ClaimId::new();
    hard.apply(&NoesisEvent::ClaimAsserted {
        claim_id: claim2_id.clone(),
        proposition: "refresh token rotates on use".into(),
        status: EpistemicStatus::Supported,
        evidence: vec![evid_id.clone()],
        scope: Scope::global("rivet", hard.revision),
        timestamp: chrono::Utc::now(),
    });

    hard.apply(&NoesisEvent::ClaimContradicted {
        claim_id: claim_id.clone(),
        contradicted_by: vec![evid_id.clone()],
        reason: "regression in 401 response".into(),
        scope: Scope::global("rivet", hard.revision),
        timestamp: chrono::Utc::now(),
    });

    let files = vec!["src/auth/mod.rs".to_string(), "src/main.rs".to_string()];
    let signals = vec!["high_churn".to_string(), "auth_target".to_string()];

    // 1. HYBRID Mode
    let ctx_hybrid = CompilationContext {
        hard_state: &hard,
        soft_workspace: &soft,
        goal_description: "Fix authentication regression",
        repository_id: "rivet",
        relevant_files: &files,
        repository_signals: &signals,
        token_budget: 4000,
        mode: RepresentationMode::Hybrid,
        deferred_trees_count: 3,
    };
    let payload_hybrid = CognitiveViewCompiler::compile(&ctx_hybrid);
    let rendered_hybrid = payload_hybrid.render();
    assert!(rendered_hybrid.contains("cognitive_view:"));
    assert!(rendered_hybrid.contains("Fix authentication regression"));
    assert!(rendered_hybrid.contains("H1: OAuth token expiry mismatch"));
    assert!(rendered_hybrid.contains("contradictions:"));
    assert!(rendered_hybrid.contains("regression in 401 response"));
    assert!(rendered_hybrid.contains("deferred_trees: 3"));

    // 2. RAW_TEXT Mode
    let ctx_raw = CompilationContext {
        mode: RepresentationMode::RawText,
        ..ctx_hybrid
    };
    let payload_raw = CognitiveViewCompiler::compile(&ctx_raw);
    let rendered_raw = payload_raw.render();
    assert!(rendered_raw.contains("=== COGNITIVE STATE (RAW TEXT) ==="));
    assert!(rendered_raw.contains("Goal: Fix authentication regression"));
    assert!(rendered_raw.contains("Contradictions"));

    // 3. TRIPLES Mode
    let ctx_triples = CompilationContext {
        mode: RepresentationMode::Triples,
        ..ctx_hybrid
    };
    let payload_triples = CognitiveViewCompiler::compile(&ctx_triples);
    let rendered_triples = payload_triples.render();
    assert!(rendered_triples.contains("# COGNITIVE VIEW (TRIPLES MODE - KNOWLEDGE GRAPH)"));
    assert!(rendered_triples.contains("requires_obligation"));
    assert!(rendered_triples.contains("epistemic_status"));

    // 4. PATHS Mode
    let ctx_paths = CompilationContext {
        mode: RepresentationMode::Paths,
        ..ctx_hybrid
    };
    let payload_paths = CognitiveViewCompiler::compile(&ctx_paths);
    let rendered_paths = payload_paths.render();
    assert!(rendered_paths.contains("# COGNITIVE VIEW (PATHS MODE - PROVENANCE GRAPH)"));
    assert!(rendered_paths.contains("Goal(Fix authentication regression) -> Obligation("));
}

#[test]
fn test_tiktoken_bpe_token_counter() {
    use rivet_view::TokenCounter;

    let sample_prompt = "Fix database concurrency bug in src/db.rs and ensure cargo test passes";
    let token_count = TokenCounter::count_tokens(sample_prompt);
    assert!(token_count > 5);
    assert!(token_count < 30);

    let empty_count = TokenCounter::count_tokens("");
    assert_eq!(empty_count, 0);
}

File: packages/core/test/rivet/view-compiler.test.ts
import { describe, expect, test } from "bun:test"
import {
  CognitiveViewCompiler,
  type CompilationContext,
} from "../../src/rivet/view-compiler"
import { HardState, SoftWorkspace } from "../../src/rivet/noesis"
import {
  Revision,
  Scope,
  createClaimId,
  createEvidenceId,
  createObligationId,
  createSessionId,
} from "../../src/rivet/types"

describe("Cognitive View Compiler Pipeline & Modes", () => {
  test("Compiles and renders all 4 representation modes: RAW_TEXT, TRIPLES, PATHS, HYBRID", () => {
    const hardState = new HardState()
    const claimId = createClaimId()
    const obligationId = createObligationId()
    const evidenceId = createEvidenceId()

    hardState.apply({
      type: "evidence_recorded",
      evidenceId,
      source: "test_output",
      summary: "cargo test pass",
      timestamp: new Date().toISOString(),
    })

    hardState.apply({
      type: "claim_asserted",
      claimId,
      proposition: "System compiles without errors",
      status: "supported",
      evidence: [evidenceId],
      scope: Scope.global("rivet", Revision.ZERO),
      timestamp: new Date().toISOString(),
    })

    hardState.apply({
      type: "obligation_created",
      obligationId,
      description: "Must verify unit tests",
      scope: Scope.global("rivet", Revision.ZERO),
      timestamp: new Date().toISOString(),
    })

    const softWorkspace = new SoftWorkspace(createSessionId(), Revision.ZERO)
    softWorkspace.setFocus(["compiler optimization"])
    softWorkspace.addHypothesis("AST representation speeds up analysis")

    const modes: ("RAW_TEXT" | "TRIPLES" | "PATHS" | "HYBRID")[] = [
      "RAW_TEXT",
      "TRIPLES",
      "PATHS",
      "HYBRID",
    ]

    for (const mode of modes) {
      const ctx: CompilationContext = {
        hardState,
        softWorkspace,
        goalDescription: "Verify compiler pipeline",
        repositoryId: "rivet",
        relevantFiles: ["src/lib.ts"],
        repositorySignals: ["typescript: 100%"],
        tokenBudget: 2000,
        mode,
        deferredTreesCount: 2,
      }

      const compiled = CognitiveViewCompiler.compile(ctx)
      const rendered = CognitiveViewCompiler.render(compiled)

      expect(typeof rendered).toBe("string")
      expect(rendered.length).toBeGreaterThan(0)

      if (mode === "RAW_TEXT") {
        expect(rendered).toContain("=== COGNITIVE STATE (RAW TEXT) ===")
        expect(rendered).toContain("System compiles without errors")
      } else if (mode === "TRIPLES") {
        expect(rendered).toContain("TRIPLES MODE")
        expect(rendered).toContain("epistemic_status")
      } else if (mode === "PATHS") {
        expect(rendered).toContain("PATHS MODE")
      } else if (mode === "HYBRID") {
        expect(rendered).toContain("```yaml")
        expect(rendered).toContain("hard_revision")
      }
    }
  })
})
